BB-338 Suspend Lifecycle transition to paused location

welansari opened the pull request:
Issue: [BB-338](https://scality.atlassian.net/browse/BB-338)

*Changes :*

- Adapt the same Pause/Resume API routes used for replication and ingestion to also support lifecycle routes
- Lifecycle routes will use mongo to store the location status
- Backbeat API will be the only process modifying the state of locations in MongoDB, hence scheduling a resume job will be handled in Backbeat API contrary to how it works in other extensions.
- The lifecycle bucket processor will be listening to mongo events using change streams and will be pausing and resuming locations based on the modifications to the state applied to MongoDB

**Note for reviewers:** 
Please review per commit as it will make understanding the changes easier.

[BB-338]: https://scality.atlassian.net/browse/BB-338?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-338 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.4`


Please check the `Fix Version/s` of BB-338, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _setMetastoreChangeStream(cb) {
```
this is not the "metastore" collection (i.e. the collection with the list of buckets), but the "location status" collection

francoisferrand commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -340,6 +508,9 @@ class LifecycleBucketProcessor {
             done => this._initKafkaBacklogMetrics(done),
             done => this._setupConsumer(done),
+            done => this._setupMongoClient(done),
+            done => this._initializeLocationStatuses(done),
```
there is a race condition here:
* Read initial state → paused
* User triggers a resume, which is added to mongo & all...
* Create the change stream → no event received
* Processor is still paused...

we should first setup the change stream, then look for the initial state ; and in the change stream callback we need to be resilient to both:
* being the first event, e.g. happening before we could do the initial state management
* being already in the expected state
(should not be hard, but just to keep in ming)

francoisferrand commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
+     * @param {function} cb callback
+     * @returns {undefined} undefined
+     */
+    _setupMongoClient(cb) {
```
it seems to me the whole mongodb change stream / status management could easily be extracted and made reusable : just need to eventually be able to call "resume" and "suspend" functions on the processor (possibly via callbacks)

it will then be easier to port this change to CRR and Ingestion....

francoisferrand commented on extensions/lifecycle/bucketProcessor/LifecycleBucketProcessor.js:
```diff
@@ -328,6 +373,129 @@ class LifecycleBucketProcessor {
+            // creating a collection in case it doesn't exist, this is done as a safety measure
+            // but normally it will be created by backbeat API that gets deployed before the
+            // lifecycle bucket processor.
+            return this._mongoClient.createCollection(locationStatusCollection, err => {
```
> locationStatusCollection

don't we need to add "filters" in other parts of Backbeat, to ensure this collection is not unexpectedly processed by replication/ingestion/... ?

welansari replied:
I think there's no need for that, in change stream implementation we only listen to buckets defined on the metastore, and in the oplog implementation, this would still be ignored as we only replicate proper objects that have a replication status.

francoisferrand commented on lib/api/constants.js:
```diff
@@ -0,0 +1,5 @@
+const constants = {
+    locationStatusCollection: '__locationStatusStore',
```
does not seem related to "api", i would expect to find this closer to database : is there some "central" place where we already define the name of the metastore or other system collections?

welansari replied:
We don't have anything like that actually, the `metastore` collection name for example is hardcoded into the mongo client in arsenal

francoisferrand commented on lib/api/BackbeatAPI.js:
```diff
@@ -1088,22 +1124,27 @@ class BackbeatAPI {
-        sites.forEach(site => {
+        const sites = this.getRequestedSites(details);
+        async.eachLimit(sites, 10, (site, next) => {
+            if (service === 'lifecycle') {
```
this does not look really nice, with this state management intermingled everywhere (which was indeed already the case)

I am wondering if it would not be better to introduce some kind of interface "LocationStatusManager", called by this api : so we could have 2 implementations (redis/zookeeper and mongo), and easily switch one or the other...

That would be much cleaner IMO, but i am wondering if this is worth it (and how long that would take), esp. since we should eventually change this for CRR and ingestion as well... On the positiive side, it would make it easier to keep both codes around, progressively migrate (e.g. just ingestion) and possibly keep the code compatible with RING (where we will not have mongo, in case we want to upgrade RING to backbeat 8.x)

XinLi0207 commented on lib/api/BackbeatAPI.js:
```diff
@@ -1171,7 +1261,18 @@ class BackbeatAPI {
 
-        sites.forEach(site => {
+        return async.eachLimit(sites, 10, (site, next) => {
+            if (service === 'lifecycle') {
```
we have three services here and doing different kinds of things for each, just my personal thinking, maybe a `switch(service) case` logic is more clear?

Commit:
BB-338 add pause/resume API for lifecycle locations

Added special class to handle the states of different services of a location
it supports both the old implementation with zookeeper and redis, and the new
one with mongo

Commit:
BB-338 support pause/resume of transition to locations

The lifecycle bucket processor watches the mongo collection where
location pause statuses is stored via change streams, and updates
the pauses status of each location.

nicolas2bert commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -916,6 +918,11 @@ class LifecycleTask extends BackbeatTask {
                 }),
             (objectMD, next) => {
+                // No transition if lifecycle events to location were paused
+                if (this.pausedLocations.has(params.site)) {
```
*perf* What about moving it up as the "first task" to avoid extra GET requests (from _getObjectMD)?

nicolas2bert commented on extensions/lifecycle/tasks/LifecycleTask.js:
```diff
@@ -916,6 +918,11 @@ class LifecycleTask extends BackbeatTask {
             (objectMD, next) => {
+                // No transition if lifecycle events to location were paused
+                if (this.pausedLocations.has(params.site)) {
+                    return next(errorLocationPaused);
```
*debug* Should we add a log with the paused location name?

welansari replied:
we already have that in https://github.com/scality/backbeat/blob/375383c1b4651629349b550784cdcb3bdbd8369d/extensions/lifecycle/tasks/LifecycleTask.js#L976-L983

nicolas2bert commented on extensions/utils/LocationStatusStream.js:
```diff
@@ -0,0 +1,180 @@
+/**
+ * @class LocationStatusStream
+ *
+ * @classdesc uses change streams to listen to a location statuss
```
*nit* status

nicolas2bert commented on extensions/utils/LocationStatusStream.js:
```diff
@@ -0,0 +1,180 @@
+            // creating a collection in case it doesn't exist, this is done as a safety measure
+            // but normally it will be created by backbeat API that gets deployed before the
+            // lifecycle bucket processor.
+            return this._mongoClient.createCollection(locationStatusCollection, err => {
```
*nit* No need to explicitly create the MongoDB collection unless specific options/config are needed.

nicolas2bert commented on extensions/utils/LocationStatusStream.js:
```diff
@@ -0,0 +1,180 @@
+                // in case the collection already exists, we ignore the error
+                if (err && err.codeName !== 'NamespaceExists') {
+                    this._log.error('Could not create mongo collection', {
+                        collection: locationStatusCollection,
```
*debug* You could add  the method to this log.

nicolas2bert commented on extensions/utils/LocationStatusStream.js:
```diff
@@ -0,0 +1,180 @@
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _setChangeStream(cb) {
```
*nit* This function is synchronous.

welansari replied:
The reason i used callbacks for this and all the other synchronous functions was for cleanness of the code, so that they can fit neatly into the async.series/waterfalls

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+                    error: err.message,
+                });
+            }
+            return cursor.toArray(cb);
```
same

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+        locations
+            .filter(loc => validLocationNames.includes(loc._id))
+            .forEach(loc => {
+                this._locationStatusStore[loc._id] = new LocationStatus(supportedServices, loc.value);
```
*archi* Why is `LocationStatus` in Arsenal? It seems to be "highly" correlated with Backbeat.

welansari replied:
The initial reason was cause it's a MongoDB document model, which means that it can get accessed by everyone.

But i guess currently the information is only accessed by backbeat and the logic of pause/resume is also very specific to backbeat. I moved the model to backbeat

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+     * @param {function} cb callback
+     * @returns {undefined}
+     */
+    _getPreviousLocationStates(locations, cb) {
```
*nit* This function looks synchronous.

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+            }
+        }, err => {
+            if (err) {
+                this._logger.error('Couldn not delete invalid locations', {
```
*nit* Could

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+                };
+            }
+            return { hours: parseInt(reqBody.hours, 10) };
+        } catch (e) {
```
*debug* should we log the error?

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+                return this._updateServiceStatusForLocation(location, next);
+            }
+            return next();
+        }, () => {
```
What if `_updateServiceStatusForLocation` fails, should we return an error to our user?
Also, might want to make sure it returns an "arsenal type of error".

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+                    return this._updateServiceStatusForLocation(location, next);
+                }
+            return next();
+        }, () => {
```
same for error.

nicolas2bert commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,552 @@
+                return this._updateServiceStatusForLocation(location, next);
+            }
+            return next();
+        }, () => {
```
same for error

KillianG commented on tests/unit/api/BackbeatAPI.spec.js:
```diff
@@ -5,15 +5,18 @@ const BackbeatRequest = require('../../../lib/api/BackbeatRequest');
+        .filter(site => locationConfig[site].isCold)[0];
 
-    before(() => {
+    beforeEach(() => {
```
Do we really need to create a new `BackbeatAPI` before every test ? Why wasn't it like that before ?

francoisferrand commented on lib/api/BackbeatAPI.js:
```diff
@@ -1383,14 +1253,47 @@ class BackbeatAPI {
+                this._mongoClient,
+                this._zkClient,
+                this._redisPublisher,
+                {
+                    crr: this._crrTopic,
+                    ingestion: this._ingestionTopic,
+                },
+                {
+                    crr: {
+                        namespace: zookeeperReplicationNamespace,
+                        statePath: zkReplicationStatePath,
+                    },
+                    ingestion: {
+                        namespace: zookeeperIngestionNamespace,
+                        statePath: zkIngestionStatePath,
+                    },
+                },
```
nit: maybe we should merge the "topics" and "zookeeper" params in a single map:

```suggestion
                {
                    crr: {
                        topic: this._crrTopic,
                        namespace: zookeeperReplicationNamespace,
                        statePath: zkReplicationStatePath,
                    },
                    ingestion: {
                        topic: this._ingestionTopic,
                        namespace: zookeeperIngestionNamespace,
                        statePath: zkIngestionStatePath,
                    },
                },
```

francoisferrand replied:
and also merge the information from `mongoLocationStatusServices`, somehow: e.g. should have an entry to the service even if it is mongo, though in that case it would NOT have the redis topic or zookeeper state.

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,583 @@
+        this._locationStatusColl = null;
+
+        this._locationStatusStore = {};
+        this._scheduledResumeJobs = {
```
can this/should this be an attribute of `LocationStatus` ?
(and thus avaialble through the `locationStatusStore` map)

welansari replied:
`_scheduledResumeJobs` contains the actual scheduled jobs for all services (that use both mongo and zookeeper), this can't be put in the `LocationStatus` as it'll only work for mongo services

francoisferrand commented on lib/util/LocationStatusManager.js:
```diff
@@ -0,0 +1,583 @@
+    _getStateDetails(service, locations, cb) {
+        if (this._mongoServices.includes(service)) {
+            const selectedLocations = {};
+            locations.forEach(loc => {
+                selectedLocations[loc] = this._locationStatusStore[loc];
+            });
```
this should do the mapping done in the later functions (`.paused = .getServicePauseStatus(service)` etc, c.f. lines 315, 342...), to complete the abstraction :

```suggestion
            locations.forEach(loc => {
                selectedLocations[loc] = this._locationStatusStore[loc].getValue()[service]
            });
```

(or better introduce a `LocationStatus.getService()` function 😉)

francoisferrand commented on tests/unit/lib/util/LocationStatusManager.spec.js:
```diff
@@ -0,0 +1,366 @@
+        sinon.restore();
+    });
+
+    it('pauseService:: should pause selected service', done => {
```
nit: `pauseService::` should be in a describe block instead, to keep the "flow" of describing the tests

bert-e commented:
# Conflict




There is a conflict between your branch `feature/BB-338-suspend-archive` and the
destination branch `development/8.6`.

Please resolve the conflict on **the feature branch** (`feature/BB-338-suspend-archive`).

```sh
git fetch && \
git checkout origin/feature/BB-338-suspend-archive && \
git merge origin/development/8.6
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:feature/BB-338-suspend-archive
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Kerkesni commented:
/approve

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`
* `development/8.4`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`
* `development/8.4`
* `development/8.5`



Please check the status of the associated issue BB-338.

Goodbye kerkesni.
